nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs

```diff
@@ -74,8 +109,196 @@ pub enum SledEditError {
+        preexisting_dataset_ids: DatasetIdsBackfillFromDb,
+    ) -> Result<Self, SledInputError> {
+        match state {
+            SledState::Active => {
+                let inner = ActiveSledEditor::new(
+                    zones,
+                    disks,
+                    datasets,
+                    preexisting_dataset_ids,
+                )?;
+                Ok(Self(InnerSledEditor::Active(inner)))
+            }
+            SledState::Decommissioned => {
+                let inner = EditedSled {
+                    state,
+                    zones,
+                    disks,
+                    datasets,
+                    edit_counts: SledEditCounts::zeroes(),
+                };
+                Ok(Self(InnerSledEditor::Decommissioned(inner)))
+            }
+        }
+    }
+
+    pub fn for_new_active(
+        preexisting_dataset_ids: DatasetIdsBackfillFromDb,
+    ) -> Self {
+        Self(InnerSledEditor::Active(ActiveSledEditor::new_empty(
+            preexisting_dataset_ids,
+        )))
+    }
+
+    pub fn finalize(self) -> EditedSled {
+        match self.0 {
+            InnerSledEditor::Active(editor) => editor.finalize(),
+            InnerSledEditor::Decommissioned(edited) => edited,
+        }
+    }
+
+    pub fn edit_counts(&self) -> SledEditCounts {
+        match &self.0 {
+            InnerSledEditor::Active(editor) => editor.edit_counts(),
+            InnerSledEditor::Decommissioned(edited) => edited.edit_counts,
+        }
+    }
+
+    pub fn decommission(&mut self) -> Result<(), SledEditError> {
+        match &mut self.0 {
+            InnerSledEditor::Active(editor) => {
+                // Decommissioning a sled is a one-way trip that has many
+                // preconditions. We can't check all of them here (e.g., we
+                // should kick the sled out of trust quorum before
+                // decommissioning, which is entirely outside the realm of
+                // `SledEditor`. But we can do some basic checks: all of the
+                // disks, datasets, and zones for this sled should be expunged.
+                editor.validate_decommisionable()?;
+
+                // We can't take ownership of `editor` from the `&mut self`
+                // reference we have, and we need ownership to call
+                // `finalize()`. Steal the contents via `mem::swap()` with an
+                // empty editor. This isn't panic safe (i.e., if we panic
+                // between the `mem::swap()` and the reassignment to `self.0`
+                // below, we'll be left in the active state with an empty sled
+                // editor), but omicron in general is not panic safe and aborts
+                // on panic. Plus `finalize()` should never panic.
+                let mut stolen = ActiveSledEditor::new_empty(
+                    DatasetIdsBackfillFromDb::empty(),
+                );
+                mem::swap(editor, &mut stolen);
```

Comment:
Instead of all this, what about accepting an owned `SledEditor` here and returning an owned one back?  I think this means the caller would have to remove this sled editor from `sled_editors` and insert the one it got back.  But that seems okay?

Reply:
Hm, I think this makes failures super awkward? If we take `self` and return `Result<SledEditor, SledEditError>`, then on failure our caller doesn't get an editor back. I guess we could return either `(SledEditor, Result<(), SledEditError>)` or `Result<SledEditor, (SledEditor, SledEditError)>` but either of those seems worse to me than the slightly-messy internal details here.

Reply:
Fair enough!